### ISSUE TITLE
Appending quasi-quoted tokens to an existing Tokens

### ISSUE THREAD

**Issue by dtolnay** [dtolnay/quote#12]
I know this is useful but I haven't figured out the right syntax yet.

``` rust
let tokens = quote! { /* ... */ };

tokens.hypothetically_append! {
    #more #things
};
```

Workaround is:

``` rust
let mut tokens = quote! { /* ... */ };

tokens = quote! {
    #tokens
    #more #things
};
```

cc @SimonSapin

**Comment by SimonSapin** [comment]
It might be an implementation detail accidentally exposed because `macro_rules!` macros don’t have privacy, but it turns out this is possible today:

```rust
    let mut tokens = quote! { /* ... */ };

    let (more, things) = (0, 0);
    quote_each_token! { tokens
        #more #things
    };
```

**Comment by dtolnay** [comment]
That's not supposed to be public and don't use it because it may break in a minor version. Something similar may be the right solution though:

```rust
let mut tokens = /* ... */;

append_quote! {
    #tokens
    #more things
};
```

I guess this differs from the "workaround" above in performance only.

**Comment by SimonSapin** [comment]
Right. And as I mentioned on IRC, I only looked into how to do this out of habit of avoiding memory allocations when possible, but then decided it’s not that important for what I was doing: code running at compile-time with a fixed amount of data that is small enough that the whole thing is fast enough already.

**Comment by colin-kiegel** [comment]
I currently find myself doing things like
```rust
let mut tokens = /* ... */;
tokens.append(&quote!(/* ... */).to_string()); // <-- (*)
```
I am already considering to define a macro `append_quote!(tokens, /* ... */)` that expands to the line marked with (*) above.

`quote` could export `append_quote!` as a thin wrapper around `quote_each_token!`? The entry point of `append_quote` could even do a [type check](https://play.rust-lang.org/?gist=4ae8a75d5b169761d126f073a35fee28&version=stable&backtrace=0) like `let $tokens: $crate::Tokens = $tokens`.

IMO that would be really nice. What do you think? Would you perhaps accept a PR?

**Comment by dtolnay** [comment]
How do you feel about `append_quote!(tokens, /* ... */)` in comparison to:

```rust
tokens.append(quote! {
    /* ... */
});
```

If we could make that work, would you still need a macro? The words are the same but the punctuation has been shuffled around in a way that I find more intuitive.

I would accept a PR that makes `append` take a trait that is implemented for &str and Tokens.

**Comment by colin-kiegel** [comment]
That's definitely more ergonomic, but at the cost of one additional allocation. I could live with that. :-)

We could `impl AsRef<str> for Tokens`. That way `append` could accept `&str`, `String` and `Tokens` at once and don't need to invent a new trait.

**Comment by colin-kiegel** [comment]
PS: Or maybe do both?

The additional macro could also be called `quote_append!(token, /* ... */)`. I think that word ordering is better and it would just be a thin wrapper macro.

**Comment by dtolnay** [comment]
I would accept `quote_append!(token, /* ... */)` if somebody can demonstrate that it gives >0.5% compile time improvement in a real project. My experience has been that proc macro performance is not worth worrying about.

**Comment by dtolnay** [comment]
Addressed in #29.
